### PR TITLE
[SHUFFLE] [WIP] Prototype: store shuffle file on external storage like S3

### PR DESCRIPTION
**Pull request by hiboyang**
### What changes were proposed in this pull request?
This PR ([design doc](https://docs.google.com/document/d/10rhvjXUlbQfWg-zh02_aqRqDT_ZnwYmICPAR--aRv64)) provides support to store shuffle files on external shuffle storage like S3. It helps Dynamic
Allocation on Kubernetes. Spark driver could release idle executors without worrying about losing
shuffle data because the shuffle data is store on external shuffle storage which are different
from executors.

This could be viewed as a followup work for https://issues.apache.org/jira/browse/SPARK-25299.

There is previously Worker Decommission feature ([SPARK-33545](https://issues.apache.org/jira/browse/SPARK-33545)), which is a great feature to copy shuffle data to fallback storage like S3. People appreciate that work to address the critical issue to handle shuffle data on Spark executor termination. The work in the PR does not intent to replace that feature. The intent is to get further discussion about how to save shuffle data on S3 during normal execution time.

### Why are the changes needed?

To better support Dynamic Allocation on Kubernetes, we need to decouple shuffle data from Spark
executor. This PR implements another Shuffle Manager and support writing shuffle data on S3.

### Does this PR introduce _any_ user-facing change?

Yes, this PR adds two Spark config like following to plug in another StarShuffleManager and store
shuffle data on provided S3 location.
```
spark.shuffle.manager=org.apache.spark.shuffle.StarShuffleManager
spark.shuffle.star.rootDir=s3://my_bucket_name/my_shuffle_folder
```

### How was this patch tested?

Added a unit test for StartShuffleManager. A lot of classes are copied from Spark, thus not add tests
for those classes. We will work with the community to get feedback first, then work on removing code
copy/duplication.

### CODE REVIEW

**Comment by c21** [comment]
@hiboyang, thanks for the work here! Could you create a design doc for this? That might help get more people attention and easier for them to understand.

**Comment by hiboyang** [comment]
> @hiboyang, thanks for the work here! Could you create a design doc for this? That might help get more people attention and easier for them to understand.

Yes, good suggestion. Will create some design doc.

**Comment by linzebing** [comment]
Quickly glanced through the code, seems for writing shuffle data we are writing locally first and then upload to S3, similarly for reading shuffle data we are downloading data to a local temp file first and then read.

We should be able to write/read direct to/from S3, right?

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Hi, All. Thank you!

BTW, for the record, Apache Spark 3.1+ already stores its shuffle files into the external storage like S3 and reads back from it.
  - [SPARK-33545][CORE] Support Fallback Storage during Worker decommission (Apache Spark 3.1.0)
  - [SPARK-34142][CORE] Support Fallback Storage Cleanup during stopping SparkContext (Apache Spark 3.2.0)
  - [SPARK-37509][CORE] Improve Fallback Storage upload speed by avoiding S3 rate limiter (Apache Spark 3.3.0)

It would be great not to ignore the existing Spark feature and avoid over-claiming.

Dynamic allocation is the same. Apache Spark has been supporting Dynamic Allocation in K8s too.

**Comment by hiboyang** [comment]
> Quickly glanced through the code, seems for writing shuffle data we are writing locally first and then upload to S3, similarly for reading shuffle data we are downloading data to a local temp file first and then read.
> 
> We should be able to write/read direct to/from S3, right?

Thanks for looking! Yes, we should be able to write/read direct on S3. This PR is a prototype. Still need to improve the code and performance of writing/reading shuffle data on S3.

**Comment by hiboyang** [comment]
> Hi, All. Thank you!
> 
> BTW, for the record, Apache Spark 3.1+ already stores its shuffle files into the external storage like S3 and reads back from it.
> 
> * [[SPARK-33545](https://issues.apache.org/jira/browse/SPARK-33545)][CORE] Support Fallback Storage during Worker decommission (Apache Spark 3.1.0)
> * [[SPARK-34142](https://issues.apache.org/jira/browse/SPARK-34142)][CORE] Support Fallback Storage Cleanup during stopping SparkContext (Apache Spark 3.2.0)
> * [[SPARK-37509](https://issues.apache.org/jira/browse/SPARK-37509)][CORE] Improve Fallback Storage upload speed by avoiding S3 rate limiter (Apache Spark 3.3.0)
> 
> It would be great not to ignore the existing Spark feature and avoid over-claiming.
> 
> Dynamic allocation is the same. Apache Spark has been supporting Dynamic Allocation in K8s too.

Right, Spark has shuffle tracking to support Dynamic Allocation on Kubernetes, but it will not work well when there is shuffle data distributed on many executors (those executors cannot be released).

The work here (storing shuffle data on S3) does not conflict with worker decommission feature. The eventual goal is to store shuffle data on S3 or other external storage directly. Before getting there, people could still use the worker decommission feature.

**Comment by dongjoon-hyun** [comment]
You are completely wrong because you already know the worker decommission feature.
>  but it will not work well when there is shuffle data distributed on many executors (those executors cannot be released).

You should mention this in the PR description explicitly instead of misleading the users.
> The work here (storing shuffle data on S3) does not conflict with worker decommission feature. The eventual goal is to store shuffle data on S3 or other external storage directly.

**Comment by hiboyang** [comment]
Add a [design doc](https://docs.google.com/document/d/10rhvjXUlbQfWg-zh02_aqRqDT_ZnwYmICPAR--aRv64) for this prototype.

**Review by steveloughran — COMMENTED** [review]
some quick comments on the pr

**Comment by steveloughran** [comment]
Obviously I am biased, but I believe that rather than trying to use the AWS APIs yourself, you should just use the hadoop file system APIs and interact with S3 through the s3a connector.

For a high-performance upload of a local file, use `FileSystem.copyFromLocalFile` in s3a on hadoop 3.3.2 this uses the same transfer manager class as this PR does but adds: exception handling/mapping, encryption settings, auditing. And the s3a integration tests verify all this works... By the time you get to use it here you can assume the S3 upload works, and it becomes a matter of linking it up to spark.


As `copyFromLocalFile` is implemented for all filesystems, it means the component will also work with other stores including google cloud and azure abfs, even if they do not override the base method for a high-performance implementation -yet.

This also means that you could write tests for the feature using file:// as the destination store and include these in the spark module; if you design such tests to be overrideable to work with other file systems, they could be picked up and reused as the actual integration test suites in an external module.

And, because someone else owns the problem of the s3 connector binding, you get to avoid fielding support calls related to configuring of AWS endpoint, region, support for third-party s3 stores, qualifying AWS SDK updates, etc.

Accordingly, I would propose 

* ripping out `StarS3ShuffleFileManager` and using FileSystem APIs
* write tests which can be pointed at the local FS, but then targeted at real object stores in a downstream test module, as https://github.com/hortonworks-spark/cloud-integration does for committer testing.

Getting integration tests set up is inevitably going to be somewhat complicated. I can provide a bit of consultation there.

File: external-shuffle-storage/pom.xml [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        <dependency>
+            <groupId>commons-io</groupId>
+            <artifactId>commons-io</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.amazonaws</groupId>
```
[the remark this answers is not in the document]
Comment (hiboyang) [reply external-shuffle-storage/pom.xml:103]: Thanks for the suggestion! Yes, I was thinking to use that hadoop library as well, then did not do it due to wanting to start small with this prototype. It sounds a good idea to switch to hadoop library.

File: external-shuffle-storage/src/main/java/org/apache/spark/starshuffle/StarBlockStoreClient.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                        if (byteBuffer.position() == size) {
+                            break;
+                        }
+                        b = inputStream.read();
+                    }
+                    byteBuffer.flip();
+                    NioManagedBuffer managedBuffer = new NioManagedBuffer(byteBuffer);
+                    listener.onBlockFetchSuccess(blockIdStr, managedBuffer);
+                } catch (IOException e) {
+                    throw new RuntimeException(String.format(
```
[the remark this answers is not in the document]
Comment (hiboyang) [reply external-shuffle-storage/src/main/java/org/apache/spark/starshuffle/StarBlockStoreClient.java:114]: Yes, good catch! Will add inner excepion!

File: external-shuffle-storage/src/main/java/org/apache/spark/starshuffle/StarBypassMergeSortShuffleWriter.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.None$;
+import scala.Option;
+import scala.Product2;
+import scala.Tuple2;
+import scala.collection.Iterator;
+
+import javax.annotation.Nullable;
+import java.io.*;
```
[the remark this answers is not in the document]
Comment (hiboyang) [reply external-shuffle-storage/src/main/java/org/apache/spark/starshuffle/StarBypassMergeSortShuffleWriter.java:45]: Good point, let me remove .* here.

File: external-shuffle-storage/src/main/java/org/apache/spark/starshuffle/StarS3ShuffleFileManager.java [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+    public static final String MAX_THREADS = "fs.s3a.threads.max";
+    public static final int DEFAULT_MAX_THREADS = 10;
+    public static final String KEEPALIVE_TIME = "fs.s3a.threads.keepalivetime";
+    public static final int DEFAULT_KEEPALIVE_TIME = 60;
+
+    public static final String AWS_REGION = "fs.s3a.endpoint.region";
+    public static final String DEFAULT_AWS_REGION = Regions.US_WEST_2.getName();
+
+    private static TransferManager transferManager;
+    private static Object transferManagerLock = new Object();
```
[the remark this answers is not in the document]
Comment (hiboyang) [reply external-shuffle-storage/src/main/java/org/apache/spark/starshuffle/StarS3ShuffleFileManager.java:67]: yes!

File: external-shuffle-storage/src/main/java/org/apache/spark/starshuffle/StarS3ShuffleFileManager.java [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+            if (duration != 0) {
+                mbs = ((double) size) / (1000 * 1000) / ((double) duration / 1000);
+            }
+            logger.info("S3 download finished: {}, file size: {} bytes, total transferred: {}, throughput: {} mbs",
+                    s3Url, size, totalTransferredBytes.get(), mbs);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(String.format(
+                    "Failed to download shuffle file %s", s3Url));
+        } finally {
+            transferManager.shutdownNow();
```
[the remark this answers is not in the document]
Comment (hiboyang) [reply external-shuffle-storage/src/main/java/org/apache/spark/starshuffle/StarS3ShuffleFileManager.java:215]: This was a code mistake. I should not shutdown shuffle manager here. Will remove this.

**Comment by hiboyang** [comment]
> Obviously I am biased, but I believe that rather than trying to use the AWS APIs yourself, you should just use the hadoop file system APIs and interact with S3 through the s3a connector.
> 
> For a high-performance upload of a local file, use `FileSystem.copyFromLocalFile` in s3a on hadoop 3.3.2 this uses the same transfer manager class as this PR does but adds: exception handling/mapping, encryption settings, auditing. And the s3a integration tests verify all this works... By the time you get to use it here you can assume the S3 upload works, and it becomes a matter of linking it up to spark.
> 
> As `copyFromLocalFile` is implemented for all filesystems, it means the component will also work with other stores including google cloud and azure abfs, even if they do not override the base method for a high-performance implementation -yet.
> 
> This also means that you could write tests for the feature using file:// as the destination store and include these in the spark module; if you design such tests to be overrideable to work with other file systems, they could be picked up and reused as the actual integration test suites in an external module.
> 
> And, because someone else owns the problem of the s3 connector binding, you get to avoid fielding support calls related to configuring of AWS endpoint, region, support for third-party s3 stores, qualifying AWS SDK updates, etc.
> 
> Accordingly, I would propose
> 
> * ripping out `StarS3ShuffleFileManager` and using FileSystem APIs
> * write tests which can be pointed at the local FS, but then targeted at real object stores in a downstream test module, as https://github.com/hortonworks-spark/cloud-integration does for committer testing.
> 
> Getting integration tests set up is inevitably going to be somewhat complicated. I can provide a bit of consultation there.

Yes, these are great suggestions! Thanks again! I will find time to make change for this, and may also reach out to your for consultation when adding integration test :)

**Comment by pspoerri** [comment]
@hiboyang I looked at your work earlier this year and I wanted to let you know that I used it as a basis for a shuffle plugin. Ultimately I decided to rewrite the plugin from scratch (except the tests) and orient myself on the design of the Spark shuffle manager.

The code is available here: https://github.com/ibm/spark-s3-shuffle/ . It acts as an external Spark plugin and allows to be loaded into Spark binary releases as a plugin.

I'm open to contributing this work back to Apache Spark if there is any interest.

**Comment by dongjoon-hyun** [comment]
Apache Spark community is open for any contribution, @pspoerri . You can make your own PR.
BTW, do you have any sharable result in terms of the stability and performance, @pspoerri ?

**Comment by hiboyang** [comment]
Hi @pspoerri great you are working on this and thanks for letting us know! I stopped working on my previous PR due to changing of work priority, but still would like to see people continue working in this area.

There is big value if storing Spark shuffle data on S3. It will save cost and also make Spark more resilient to disk error.

During my previous experiment, shuffle data on S3 will have very worse performance. There are a lot of optimization needed, e.g. S3 key prefix randomization to avoid S3 throttling and async S3 write. Will be happy to hear your thoughts on this as well.

**Comment by steveloughran** [comment]
@michaelbilow hadoop s3a is on v2 sdk; the com.amazonaws classes are not on the CP and amazon are slowly stopping support. you cannot for example use the lower latency S3 express stores with it.

Like I say: I think you would be better off using the Hue file system APIs to talk to s3. If there are aspects of s3 storage which aren't available through the API -or just very inefficiently due to the effort to preserve the Posix metaphor, then lets fix the API so that other stores can offer the same features, and other apps can pick up.

For example, here's our ongoing delete API for iceberg and other manifest-based tables
https://github.com/apache/hadoop/pull/6726
It maps to s3 bulk delete calls, but there's scope to add to other stores (we now actually want to add it as a page-size == 1 option for all filesystems as it simplifies iceberg integration).

**Comment by pspoerri** [comment]
@steveloughran How do I call the Hue APIs from Spark? Can you point me to a package?
I agree with you that using the Hadoop APIs are not ideal performance wise, but they are great from a usability and portability perspective.

Another issue is that Hadoop wants to know the size of every file it wants to read. While this makes sense for formats like parquet where the header is located at the last few bytes of the file. It does not make sense for shuffle where you know the exact block/file you want to read.

**Comment by xleoken** [comment]
+1, would be great to get this working and part of the Spark ecosystem.

**Pull request closed without merging**